napari/_qt/_qapp_model/qactions/_layer.py

```diff
@@ -0,0 +1,187 @@
+                loaded_attr_value = np.array(loaded_attr_value)
+            if key == 'shear':
+                loaded_attr_value = loaded_attr_value[
+                    -(layer.ndim * (layer.ndim - 1)) // 2 :
+                ]
+            elif key == 'affine':
+                loaded_attr_value = loaded_attr_value[
+                    -(layer.ndim + 1) :, -(layer.ndim + 1) :
+                ]
+            elif isinstance(loaded_attr_value, np.ndarray):
+                if loaded_attr_value.ndim == 1:
+                    loaded_attr_value = loaded_attr_value[-layer.ndim :]
+                elif loaded_attr_value.ndim == 2:
+                    loaded_attr_value = loaded_attr_value[
+                        -layer.ndim :, -layer.ndim :
+                    ]
+
+            setattr(layer, key, loaded_attr_value)
+
+
+def is_valid_spatial_in_clipboard() -> bool:
+    try:
+        loaded = _get_spatial_from_clipboard()
+    except (json.JSONDecodeError, pickle.UnpicklingError):
+        return False
+    if not isinstance(loaded, dict):
+        return False
+
+    return set(loaded).issubset(
+        {'affine', 'rotate', 'scale', 'shear', 'translate'}
+    )
+
+
+Q_LAYER_ACTIONS = [
+    Action(
+        id='napari.layer.copy_all_to_clipboard',
+        title=trans._('Copy all to clipboard'),
+        callback=_copy_spatial_to_clipboard,
+        menus=[{'id': MenuId.LAYERS_COPY_SPATIAL}],
+    ),
+    Action(
+        id='napari.layer.copy_affine_to_clipboard',
+        title=trans._('Copy affine to clipboard'),
+        callback=_copy_affine_to_clipboard,
+        menus=[{'id': MenuId.LAYERS_COPY_SPATIAL}],
+    ),
+    Action(
+        id='napari.layer.copy_rotate_to_clipboard',
+        title=trans._('Copy rotate to clipboard'),
+        callback=_copy_rotate_to_clipboard,
+        menus=[{'id': MenuId.LAYERS_COPY_SPATIAL}],
+    ),
+    Action(
+        id='napari.layer.copy_scale_to_clipboard',
+        title=trans._('Copy scale to clipboard'),
+        callback=_copy_scale_to_clipboard,
+        menus=[{'id': MenuId.LAYERS_COPY_SPATIAL}],
+    ),
+    Action(
+        id='napari.layer.copy_shear_to_clipboard',
+        title=trans._('Copy shear to clipboard'),
+        callback=_copy_shear_to_clipboard,
+        menus=[{'id': MenuId.LAYERS_COPY_SPATIAL}],
+    ),
+    Action(
+        id='napari.layer.copy_translate_to_clipboard',
+        title=trans._('Copy translate to clipboard'),
+        callback=_copy_translate_to_clipboard,
+        menus=[{'id': MenuId.LAYERS_COPY_SPATIAL}],
+    ),
+    Action(
+        id='napari.layer.paste_spatial_from_clipboard',
+        title=trans._('Apply scale/transforms from Clipboard'),
+        callback=_paste_spatial_from_clipboard,
+        menus=[
+            {
+                'id': MenuId.LAYERLIST_CONTEXT,
+                'group': MenuGroup.LAYERLIST_CONTEXT.COPY_SPATIAL,
+            }
+        ],
+        enablement=parse_expression('valid_spatial_json_clipboard'),
```

Comment:
@Czaki I think just the string is fine, though the typing does specify `Expr` (I've raised an issue https://github.com/pyapp-kit/app-model/issues/196)

I think this is what happens: https://github.com/pyapp-kit/app-model/blob/8289d03f3b4fb73f886e36bc53132cc48b32d941/src/app_model/expressions/_expressions.py#L81-L83

Reply:
> @Czaki I think just the string is fine, though the typing does specify `Expr` (I've raised an issue [pyapp-kit/app-model#196](https://github.com/pyapp-kit/app-model/issues/196))

I hate type error highlight in IDE...

Reply:
This is just how pydantic works (if it can cast the object, like str, to the specified type annotation, like Expr, then it's fine).  If you want your IDE not to highlight it as an error you can add the mypy plugin for pydantic: https://docs.pydantic.dev/latest/integrations/mypy/

Reply:
I do not see how it will impact PyCharm. There is Pydantic extension for PyCharm, but it does not help.

Reply:
Yeah sure about pycharm...